runtime/src/bank.rs

```diff
@@ -4788,86 +4796,114 @@ impl Bank {
         }
     }
 
-    pub fn load_program(&self, pubkey: &Pubkey) -> Arc<LoadedProgram> {
-        let Some(program) = self.get_account_with_fixed_root(pubkey) else {
-            return Arc::new(LoadedProgram::new_tombstone(
-                self.slot,
-                LoadedProgramType::Closed,
-            ));
+    fn load_program_accounts(&self, pubkey: &Pubkey) -> ProgramAccountLoadResult {
+        let program_account = match self.get_account_with_fixed_root(pubkey) {
+            None => return ProgramAccountLoadResult::AccountNotFound,
+            Some(account) => account,
         };
-        let mut transaction_accounts = vec![(*pubkey, program)];
-        let is_upgradeable_loader =
-            bpf_loader_upgradeable::check_id(transaction_accounts[0].1.owner());
-        if is_upgradeable_loader {
-            if let Ok(UpgradeableLoaderState::Program {
-                programdata_address,
-            }) = transaction_accounts[0].1.state()
+
+        if !solana_bpf_loader_program::check_loader_id(program_account.owner()) {
+            return ProgramAccountLoadResult::InvalidAccountData;
```

Comment:
I think this originally returned `InstructionError::IncorrectProgramId` in `load_program_from_account()` and all errors from there were turned into `LoadedProgramType::FailedVerification` in `load_program()`. But it should be unreachable, so try placing a `debug_assert(solana_bpf_loader_program::check_loader_id();` here instead.